comparison/comparison.md

```diff
@@ -0,0 +1,523 @@
+# Comparison with PyTorch's DataLoader
+
+```python
+import lazy_dataset
+from torch.utils import data as torch_data
+from torch.utils.data import DataLoader as TorchDataLoader
+```
+
+| Example | Feature | `lazy_dataset` | `TorchDataLoader` |
+| ------- | ------- | :------------: | :---------------: |
+|    1    | Dataset class | `lazy_datsaset.Dataset`| `torch_data.Dataset` |
+|    2    | Batching | yes | yes |
+|    2    | Collate batch | manual | automatic |
+|    3    | Shuffle | yes | yes |
+|    4    | Multi-process data loading | yes | yes (built-in shared memory GPU) |
+|    4    | Multi-thread data loading | yes (default) | no |
+|    see [README](../README.md)    | Filter (builtins.filter) | yes | no |
+|    5    | Sort by key | yes | no |
+|    5    | Sort by value | yes | no |
+|    6    | Draw random example | yes | no |
+|    7    | Unbatch | yes | no |
+
+## Examples
+1. Dataset
+    ```python
+    >>> examples = {
+    ...    'ex_1': {
+    ...        'example_id': 'ex_1',
+    ...        'observation': [1, 2, 3],
+    ...        'label': 1
+    ...    },
+    ...    'ex_2': {
+    ...        'example_id': 'ex_2',
+    ...        'observation': [4, 5, 6],
+    ...        'label': 2
+    ...    }
+    ... }
+
+    >>> class TorchDictDataset(torch_data.Dataset):
+    ...
+    ...    def __init__(self, examples):
+    ...        self.examples = examples
+    ...
+    ...    def __len__(self):
+    ...        return len(self.examples)
+    ...
+    ...    def __getitem__(self, idx):
+    ...        idx_to_key = list(self.examples.keys())[idx]
+    ...        return self.examples[idx_to_key]
+
+    >>> torch_dict_ds = TorchDictDataset(examples)
+    >>> print(torch_dict_ds[0])
+    {'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1}
+
+    >>> dict_ds = lazy_dataset.from_dict(examples)
+    >>> print(dict_ds[0])
+    {'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1}
+    ```
+2. Batching & Collate
+    ```python
+    >>> data_loader = TorchDataLoader(
+    ...     torch_dict_ds, batch_size=2, collate_fn=None
+    ... )
+    >>> next(iter(data_loader))
+    {'example_id': ['ex_1', 'ex_2'],
+     'observation': [tensor([1, 4]), tensor([2, 5]), tensor([3, 6])],
+     'label': tensor([1, 2])}
+
+    >>> dict_ds = dict_ds.batch(2)
+    >>> next(iter(dict_ds))
+    [{'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1},
+     {'example_id': 'ex_2', 'observation': [4, 5, 6], 'label': 2}]
+    >>> def collate_fn(batch):
```

Comment:
Could you replace this function with something like 
```python
    def collate_fn(batch):
        for b in batch[1:]:
            assert batch[0].keys() == b.keys(), batch
        return {k: [b[k] for b in batch] for k in batch[0].keys()}
```
In case someone copies this code, this is more robust.

You could also add a hint, that we have in padertorch (`padertorch.data.utils.collate_fn`) a collate function that can handle nested batches.

Reply:
Updated in https://github.com/fgnt/lazy_dataset/pull/27/commits/1d9571a7ecb5867874733b2c139555b567826257